File: bot-api/dotnet/api/src/internal/BotEventHandlers.cs
using System;
using System.Collections.Generic;
using Robocode.TankRoyale.BotApi.Events;

namespace Robocode.TankRoyale.BotApi.Internal;

/// <summary>
/// Class used for bot event handlers on the public API that are may or may not be triggered bot event queue, and might
/// not be handled immediately by the bot logic.
/// </summary>
sealed class BotEventHandlers
{
    internal readonly EventHandler<ConnectedEvent> OnConnected = new();
    internal readonly EventHandler<DisconnectedEvent> OnDisconnected = new();
    internal readonly EventHandler<ConnectionErrorEvent> OnConnectionError = new();

    internal readonly EventHandler<GameStartedEvent> OnGameStarted = new();
    internal readonly EventHandler<GameEndedEvent> OnGameEnded = new();
    internal readonly EventHandler<GameAbortedEvent> OnGameAborted = new();
    internal readonly EventHandler<RoundStartedEvent> OnRoundStarted = new();
    internal readonly EventHandler<RoundEndedEvent> OnRoundEnded = new();

    private readonly EventHandler<TickEvent> _onTick = new();
    internal readonly EventHandler<SkippedTurnEvent> OnSkippedTurn = new();
    private readonly EventHandler<DeathEvent> _onDeath = new();
    private readonly EventHandler<BotDeathEvent> _onBotDeath = new();
    private readonly EventHandler<HitBotEvent> _onHitBot = new();
    private readonly EventHandler<HitWallEvent> _onHitWall = new();
    private readonly EventHandler<BulletFiredEvent> _onBulletFired = new();
    private readonly EventHandler<HitByBulletEvent> _onHitByBullet = new();
    private readonly EventHandler<BulletHitBotEvent> _onBulletHit = new();
    private readonly EventHandler<BulletHitBulletEvent> _onBulletHitBullet = new();
    private readonly EventHandler<BulletHitWallEvent> _onBulletHitWall = new();
    private readonly EventHandler<ScannedBotEvent> _onScannedBot = new();
    private readonly EventHandler<WonRoundEvent> _onWonRound = new();
    private readonly EventHandler<CustomEvent> _onCustomEvent = new();
    private readonly EventHandler<TeamMessageEvent> _onTeamMessage = new();

    private readonly Dictionary<Type, Action<IEvent>> _eventHandlers = new();


    internal BotEventHandlers(IBaseBot baseBot)
    {
        InitializeEventHandlers();
        SubscribeToEventHandlers(baseBot);
    }

    private void InitializeEventHandlers()
    {
        _eventHandlers[typeof(ConnectedEvent)] = e => OnConnected.Publish((ConnectedEvent)e);
        _eventHandlers[typeof(DisconnectedEvent)] = e => OnDisconnected.Publish((DisconnectedEvent)e);
        _eventHandlers[typeof(ConnectionErrorEvent)] = e => OnConnectionError.Publish((ConnectionErrorEvent)e);

        _eventHandlers[typeof(GameStartedEvent)] = e => OnGameStarted.Publish((GameStartedEvent)e);
        _eventHandlers[typeof(GameEndedEvent)] = e => OnGameEnded.Publish((GameEndedEvent)e);
        _eventHandlers[typeof(GameAbortedEvent)] = e => OnGameAborted.Publish((GameAbortedEvent)e);
        _eventHandlers[typeof(RoundStartedEvent)] = e => OnRoundStarted.Publish((RoundStartedEvent)e);
        _eventHandlers[typeof(RoundEndedEvent)] = e => OnRoundEnded.Publish((RoundEndedEvent)e);

        _eventHandlers[typeof(TickEvent)] = e => _onTick.Publish((TickEvent)e);
        _eventHandlers[typeof(SkippedTurnEvent)] = e => OnSkippedTurn.Publish((SkippedTurnEvent)e);
        _eventHandlers[typeof(DeathEvent)] = e => _onDeath.Publish((DeathEvent)e);
        _eventHandlers[typeof(BotDeathEvent)] = e => _onBotDeath.Publish((BotDeathEvent)e);
        _eventHandlers[typeof(HitBotEvent)] = e => _onHitBot.Publish((HitBotEvent)e);
        _eventHandlers[typeof(HitWallEvent)] = e => _onHitWall.Publish((HitWallEvent)e);
        _eventHandlers[typeof(BulletFiredEvent)] = e => _onBulletFired.Publish((BulletFiredEvent)e);
        _eventHandlers[typeof(HitByBulletEvent)] = e => _onHitByBullet.Publish((HitByBulletEvent)e);
        _eventHandlers[typeof(BulletHitBotEvent)] = e => _onBulletHit.Publish((BulletHitBotEvent)e);
        _eventHandlers[typeof(BulletHitBulletEvent)] = e => _onBulletHitBullet.Publish((BulletHitBulletEvent)e);
        _eventHandlers[typeof(BulletHitWallEvent)] = e => _onBulletHitWall.Publish((BulletHitWallEvent)e);
        _eventHandlers[typeof(ScannedBotEvent)] = e => _onScannedBot.Publish((ScannedBotEvent)e);
        _eventHandlers[typeof(WonRoundEvent)] = e => _onWonRound.Publish((WonRoundEvent)e);
        _eventHandlers[typeof(CustomEvent)] = e => _onCustomEvent.Publish((CustomEvent)e);
        _eventHandlers[typeof(TeamMessageEvent)] = e => _onTeamMessage.Publish((TeamMessageEvent)e);
    }

    private void SubscribeToEventHandlers(IBaseBot baseBot)
    {
        OnConnected.Subscribe(baseBot.OnConnected);
        OnDisconnected.Subscribe(baseBot.OnDisconnected);
        OnConnectionError.Subscribe(baseBot.OnConnectionError);

        OnGameStarted.Subscribe(baseBot.OnGameStarted);
        OnGameEnded.Subscribe(baseBot.OnGameEnded);
        OnRoundStarted.Subscribe(baseBot.OnRoundStarted);
        OnRoundEnded.Subscribe(baseBot.OnRoundEnded);

        _onTick.Subscribe(baseBot.OnTick);
        OnSkippedTurn.Subscribe(baseBot.OnSkippedTurn);
        _onDeath.Subscribe(baseBot.OnDeath);
        _onBotDeath.Subscribe(baseBot.OnBotDeath);
        _onHitBot.Subscribe(baseBot.OnHitBot);
        _onHitWall.Subscribe(baseBot.OnHitWall);
        _onBulletFired.Subscribe(baseBot.OnBulletFired);
        _onHitByBullet.Subscribe(baseBot.OnHitByBullet);
        _onBulletHit.Subscribe(baseBot.OnBulletHit);
        _onBulletHitBullet.Subscribe(baseBot.OnBulletHitBullet);
        _onBulletHitWall.Subscribe(baseBot.OnBulletHitWall);
        _onScannedBot.Subscribe(baseBot.OnScannedBot);
        _onWonRound.Subscribe(baseBot.OnWonRound);
        _onCustomEvent.Subscribe(baseBot.OnCustomEvent);
        _onTeamMessage.Subscribe(baseBot.OnTeamMessage);
    }

    internal void FireEvent(BotEvent botEvent)
    {
        if (_eventHandlers.TryGetValue(botEvent.GetType(), out var action))
        {
            action.Invoke(botEvent); // calls Publish(botEvent)
        }
        else
        {
            throw new InvalidOperationException("Unhandled event: " + botEvent);
        }
    }

    // Virtual (fake) event:
    internal sealed class GameAbortedEvent : IEvent
    {
    }
}

File: bot-api/typescript/src/events/BotEventHandlers.ts
import { EventHandler } from "./EventHandler.js";
import { BotEvent } from "./BotEvent.js";
import { TickEvent } from "./TickEvent.js";
import { ScannedBotEvent } from "./ScannedBotEvent.js";
import { HitBotEvent } from "./HitBotEvent.js";
import { HitByBulletEvent } from "./HitByBulletEvent.js";
import { HitWallEvent } from "./HitWallEvent.js";
import { BulletFiredEvent } from "./BulletFiredEvent.js";
import { BulletHitBotEvent } from "./BulletHitBotEvent.js";
import { BulletHitBulletEvent } from "./BulletHitBulletEvent.js";
import { BulletHitWallEvent } from "./BulletHitWallEvent.js";
import { BotDeathEvent } from "./BotDeathEvent.js";
import { DeathEvent } from "./DeathEvent.js";
import { SkippedTurnEvent } from "./SkippedTurnEvent.js";
import { WonRoundEvent } from "./WonRoundEvent.js";
import { CustomEvent } from "./CustomEvent.js";
import { TeamMessageEvent } from "./TeamMessageEvent.js";
import { ConnectedEvent } from "./ConnectedEvent.js";
import { DisconnectedEvent } from "./DisconnectedEvent.js";
import { ConnectionErrorEvent } from "./ConnectionErrorEvent.js";
import { GameStartedEvent } from "./GameStartedEvent.js";
import { GameEndedEvent } from "./GameEndedEvent.js";
import { RoundStartedEvent } from "./RoundStartedEvent.js";
import { RoundEndedEvent } from "./RoundEndedEvent.js";

/** Holds one EventHandler per event type and dispatches BotEvents by type. */
export class BotEventHandlers {
  readonly onTick = new EventHandler<TickEvent>();
  readonly onScannedBot = new EventHandler<ScannedBotEvent>();
  readonly onHitBot = new EventHandler<HitBotEvent>();
  readonly onHitByBullet = new EventHandler<HitByBulletEvent>();
  readonly onHitWall = new EventHandler<HitWallEvent>();
  readonly onBulletFired = new EventHandler<BulletFiredEvent>();
  readonly onBulletHitBot = new EventHandler<BulletHitBotEvent>();
  readonly onBulletHitBullet = new EventHandler<BulletHitBulletEvent>();
  readonly onBulletHitWall = new EventHandler<BulletHitWallEvent>();
  readonly onBotDeath = new EventHandler<BotDeathEvent>();
  readonly onDeath = new EventHandler<DeathEvent>();
  readonly onSkippedTurn = new EventHandler<SkippedTurnEvent>();
  readonly onWonRound = new EventHandler<WonRoundEvent>();
  readonly onCustomEvent = new EventHandler<CustomEvent>();
  readonly onTeamMessage = new EventHandler<TeamMessageEvent>();
  readonly onConnected = new EventHandler<ConnectedEvent>();
  readonly onDisconnected = new EventHandler<DisconnectedEvent>();
  readonly onConnectionError = new EventHandler<ConnectionErrorEvent>();
  readonly onGameStarted = new EventHandler<GameStartedEvent>();
  readonly onGameEnded = new EventHandler<GameEndedEvent>();
  readonly onRoundStarted = new EventHandler<RoundStartedEvent>();
  readonly onRoundEnded = new EventHandler<RoundEndedEvent>();

  fireEvent(event: BotEvent): void {
    if (event instanceof TickEvent) this.onTick.publish(event);
    else if (event instanceof ScannedBotEvent) this.onScannedBot.publish(event);
    else if (event instanceof HitBotEvent) this.onHitBot.publish(event);
    else if (event instanceof HitByBulletEvent) this.onHitByBullet.publish(event);
    else if (event instanceof HitWallEvent) this.onHitWall.publish(event);
    else if (event instanceof BulletFiredEvent) this.onBulletFired.publish(event);
    else if (event instanceof BulletHitBotEvent) this.onBulletHitBot.publish(event);
    else if (event instanceof BulletHitBulletEvent) this.onBulletHitBullet.publish(event);
    else if (event instanceof BulletHitWallEvent) this.onBulletHitWall.publish(event);
    else if (event instanceof BotDeathEvent) this.onBotDeath.publish(event);
    else if (event instanceof DeathEvent) this.onDeath.publish(event);
    else if (event instanceof SkippedTurnEvent) this.onSkippedTurn.publish(event);
    else if (event instanceof WonRoundEvent) this.onWonRound.publish(event);
    else if (event instanceof CustomEvent) this.onCustomEvent.publish(event);
    else if (event instanceof TeamMessageEvent) this.onTeamMessage.publish(event);
  }
}
